game.cc

```diff
@@ -1,49 +1,606 @@
+	for (int i = 0; i < 25; i++){
+		if (currentPlayer == WHITE){
+			if(playerWhite[i] == theBoard[endX][endY]) return false;
+		}
+		else{
+			if(playerBlack[i] == theBoard[endX][endY]) return false;
+		}
+	}
+
+	//Check for Check
+	if (checkForCheck == true){
+		if (isCheckAfterMove(startX,startY,endX,endY) == true) return false;
+	}
+
+	//Checks for Pawn Promotion,
+	if (promoteType != ' '){
+
+		//Checks that the piece is a pawn moving to the end row
+		if (theBoard[startX][startY]->getWorth() != PAWN) return false;
+		if (currentPlayer == WHITE && endY != 0) return false;
+		if (currentPlayer == BLACK && endY != 7) return false;
+
+
+
+		// Check if character is correct Case
+		if (promoteType > 'A') promoteType = promoteType - 'A' + 'a';
+
+		//Finds location of piece in the array of pieces
+		int loc = 16;
+		if (currentPlayer == WHITE){	
+			while(playerWhite[loc] != NULL){
+				++loc;
+			}
+		}
+		else{
+			while(playerBlack[loc] != NULL){
+				++loc;
+			}
+		}
+
+		if (promoteType == 'q'){
+			if (currentPlayer == WHITE){
+				playerWhite[loc] = new Queen(WHITE);
+				playerWhite[loc]->setGame(this);
+				playerWhite[loc]->setLocation(startX,startY);
+				theBoard[startX][startY] = playerWhite[loc];
+			}
+			else{
+				playerBlack[loc] = new Queen(BLACK);
+				playerBlack[loc]->setGame(this);
+				playerBlack[loc]->setLocation(startX,startY);
+				theBoard[startX][startY] = playerBlack[loc];
+			}
+		}
+		else if (promoteType == 'r'){
+			if (currentPlayer == WHITE){
+				playerWhite[loc] = new Rook(WHITE);
+				playerWhite[loc]->setGame(this);
+				playerWhite[loc]->setLocation(startX,startY);
+				theBoard[startX][startY] = playerWhite[loc];
+			}
+			else{
+				playerBlack[loc] = new Rook(BLACK);
+				playerBlack[loc]->setGame(this);
+				playerBlack[loc]->setLocation(startX,startY);
+				theBoard[startX][startY] = playerBlack[loc];
+			}
+		}
+		else if (promoteType== 'b'){
+			if (currentPlayer == WHITE){
+				playerWhite[loc] = new Bishop(WHITE);
+				playerWhite[loc]->setGame(this);
+				playerWhite[loc]->setLocation(startX,startY);
+				theBoard[startX][startY] = playerWhite[loc];
+			}
+			else{
+				playerBlack[loc] = new Bishop(BLACK);
+				playerBlack[loc]->setGame(this);
+				playerBlack[loc]->setLocation(startX,startY);
+				theBoard[startX][startY] = playerBlack[loc];
+			}
+		}
+		else if (promoteType == 'n'){
+			if (currentPlayer == WHITE){
+				playerWhite[loc] = new Knight(WHITE);
+				playerWhite[loc]->setGame(this);
+				playerWhite[loc]->setLocation(startX,startY);
+				theBoard[startX][startY] = playerWhite[loc];
+			}
+			else{
+				playerBlack[loc] = new Knight(BLACK);
+				playerBlack[loc]->setGame(this);
+				playerBlack[loc]->setLocation(startX,startY);
+				theBoard[startX][startY] = playerBlack[loc];
+			}
+		}
+	}
+
+	//Check for castle
+	if (theBoard[startX][startY]->getWorth() == KING  && (abs(startX - endX) == 2 || abs(startX - endX) == 3)){
```

Comment:
Do we need to check this if we're already doing it in the Piece itself? Code duplication?